[api-documenter] Allow customization of table of contents

Vitalius1 opened the pull request:
@octogonz here is a draft of the changes that generates a custom TOC. I tried to make them as generic as possible so that not to break anything for the existing behavior. Also, I've put some comments on some missing types that we need to define. I've verified that without a config file provided it still generates what it was before.

Let me know how else I can help!



cc: @natalieethell , @aditima , @KevinTCoughlin

natalieethell commented on apps/api-documenter/src/documenters/YamlDocumenter.ts:
```diff
@@ -668,4 +707,42 @@ export class YamlDocumenter {
+  private _generateTocPointersMap(tocConfig: any): void {
+    for (const tocItem of tocConfig.items) {
+      if (tocItem.items) {
+        if (tocItem.items.length > 0) {
```
Instead of nesting these if statements, let's do 
```tsx
if (tocItem.items && tocItem.items.length > 0) {
```

Vitalius1 replied:
Thanks... I do vaguely remember that I stopped here for a bit and had a reason why I nested them like this but it doesn't look I remember why and removing the nesting makes sense.

octogonz commented:
Could we replace the **api-documenter-oufr** test with a generic project that isolates specific interesting test cases?  Some possible concerns:

- These files are somewhat large (e.g. office-ui-fabric-react.api.json is 5MB).  This increases the `git clone` footprint, and GitHub's UI generally chokes up on large files.

- The content contains lots of keywords that are unrelated to our projects, and may reduce the quality of search results (e.g. GitHub search).

- Although your usage here is okay, in general we're very careful about copying content between GitHub projects to avoid copyright issues.  Even if both projects are owned by Microsoft, the third-party notice requirements may be different.

Also, the exercise of creating a different test may help us to think about other possible applications of your feature beyond OUIFR's requirements.

octogonz commented on apps/api-documenter/src/yaml/IYamlTocFile.ts:
```diff
@@ -34,3 +34,42 @@ export interface IYamlTocFile {
+   * that are matched with the filters provided. Everything else will be placed under a catchAll category
+   * that is highly recommended to be provided.
+   */
+  tocConfig: IYamlTocFile;
```
I feel like there should be a separation between the **toc.yml** file format (as specified by DocFX) versus the **api-documenter.json** file format (which is specific to API Documenter, and may eventually be usable with non-DocFX outputs).

Vitalius1 replied:
How do ensure then that the TOC map user provides then is somewhat aligned with DocFX specification for a TOC

Vitalius1 commented:
@octogonz I removed the test I added initially and modified the existing one by adding some inline custom tags. I've also adjusted the debugger to run it and it generates the expected TOC.

octogonz commented on apps/api-documenter/src/documenters/YamlDocumenter.ts:
```diff
@@ -490,7 +529,8 @@ export class YamlDocumenter {
     let stringified: string = yaml.safeDump(dataObject, {
-      lineWidth: 120
+      lineWidth: 120,
+      noRefs: this._config && this._config.noDuplicateEntries ? true : false
```
Why are you enabling on `noRefs`?  Generally we've shied away from this YAML feature because it introduces reference nodes that interfere with diffs and don't seem to serve any useful purpose. (A properly defined file format normally doesn't include a significant amount of redundant data.)

Vitalius1 replied:
It can be safely removed I believe as it's more of a case-specific thing I tried. In our case, we are loading 4 api.json files: OUFR, styling, utilities, and merge-styles. OUFR on its hand was also re-exporting 2 of this: styling and utilities, so I ended up having lots of duplicates in the navigation. But we will handle this differently now so no need for it now. Forgot to remove it 😄

Vitalius1 replied:
the reference nodes that you say about are actually introduced if this feature is not toggled on.

Vitalius1 commented:
@iclanton I would appreciate having this experimental addition merged and a release triggered so I can start testing it. Thanks!

iclanton commented on apps/api-documenter/.vscode/launch.json:
```diff
@@ -0,0 +1,23 @@
+            "sourceMaps": true
+          },
+    ]
+}
```
Don't check this file in.

Vitalius1 replied:
Removed 👍

iclanton commented on apps/api-documenter/src/cli/GenerateAction.ts:
```diff
@@ -0,0 +1,51 @@
+import { ApiModel } from '@microsoft/api-extractor-model';
+import { DocumenterConfig } from '../documenters/DocumenterConfig';
+import { FileSystem } from '@microsoft/node-core-library';
+import { IConfigFile } from '../documenters/IConfigFile';
```
Can you coalesce these and maybe group them by internal/external paths?

Vitalius1 replied:
Sorted 👍

iclanton commented on apps/api-documenter/src/cli/GenerateAction.ts:
```diff
@@ -0,0 +1,51 @@
+  protected onDefineParameters(): void { // override
+    super.onDefineParameters();
+  }
+
```
Remove this.

Vitalius1 replied:
Removed

iclanton commented on apps/api-documenter/src/documenters/DocumenterConfig.ts:
```diff
@@ -0,0 +1,33 @@
+   * The JSON Schema for API Extractor config file (api-extractor.schema.json).
+   */
+  public static readonly jsonSchema: JsonSchema = JsonSchema.fromFile(
+    path.join(__dirname, '../schemas/api-documenter.schema.json'));
```
>../ [](start = 26, length = 3)

Change this call to `path.resolve(__dirname, '..', 'schemas', 'api-documenter.schema.json'));

Vitalius1 replied:
Changed

iclanton commented on apps/api-documenter/src/documenters/ExperimentYamlDocumenter.ts:
```diff
@@ -0,0 +1,139 @@
+ * EXPERIMENTAL - This documenter is a prototype of a new config file driven mode of operation for
+ * API Documenter.  It is not ready for general usage yet.  Its design may change in the future.
+ */
+export class ExperimentYamlDocumenter extends YamlDocumenter {
```
>ExperimentYamlDocumenter [](start = 13, length = 24)

Rename this class to `ExperimentalYamlDocumenter`

Vitalius1 replied:
Renamed

iclanton commented on apps/api-documenter/src/documenters/ExperimentYamlDocumenter.ts:
```diff
@@ -0,0 +1,139 @@
+    super(apiModel);
+    this._config = configFile.tableOfContents!;
+
+    this._tocPointerMap = {}; // need a type?
```
>// need a type? [](start = 30, length = 15)

No, this is correct.

Vitalius1 replied:
👍

iclanton commented on apps/api-documenter/src/documenters/ExperimentYamlDocumenter.ts:
```diff
@@ -0,0 +1,139 @@
+    for (const apiItem of apiItems) {
+      let tocItem: IYamlTocItem;
+
+      if (apiItem.kind === ApiItemKind.Namespace) {
```
>if (apiItem.kind === ApiItemKind.Namespace) { [](start = 6, length = 45)

Would this be cleaner as a `switch` satement?

Vitalius1 replied:
This is currently what lives in `YamlDocumenter`. A refactor of this logic is not part of this PR, although it might be something to look into in a future PR.

iclanton replied:
That's fine.

iclanton commented on apps/api-documenter/src/documenters/YamlDocumenter.ts:
```diff
@@ -1,6 +1,8 @@
 // See LICENSE in the project root for license information.
 
+// tslint:disable:member-ordering
+
```
@octogonz - did you ever update this rule?

iclanton commented on common/changes/@microsoft/api-documenter/FabricDocumenter_2019-05-06-23-48.json:
```diff
@@ -0,0 +1,11 @@
+  "changes": [
+    {
+      "packageName": "@microsoft/api-documenter",
+      "comment": "Adds optional config file `api-documenter.json` to allow building custom TOC for YamlDocumenter.",
```
>Adds [](start = 18, length = 4)

*Add

Vitalius1 replied:
Changed

iclanton commented on common/changes/@microsoft/api-documenter/FabricDocumenter_2019-05-06-23-48.json:
```diff
@@ -0,0 +1,11 @@
+  "changes": [
+    {
+      "packageName": "@microsoft/api-documenter",
+      "comment": "Adds optional config file `api-documenter.json` to allow building custom TOC for YamlDocumenter.",
```
>TOC [](start = 91, length = 3)

Write out "Table of Contents"

Vitalius1 replied:
Changed

iclanton commented on build-tests/api-documenter-test/etc/markdown/api-documenter-test.docbaseclass.md:
```diff
@@ -6,6 +6,7 @@
 Example base class
 
+
 <b>Signature:</b>
```
Is adding this blank line intentional? Seems like it shouldn't be.

Vitalius1 replied:
Removed this blank line although it was autogenerated when I ran the test. So with future runs it might get added again.

iclanton replied:
That makes it seem like a regression was introduced.

Vitalius1 replied:
That would be highly unlikely because we did not touch the `MarkdownDocumenter` in this PR. I added a `docCategory` tag to the test files that might have somehow influenced the output.

iclanton commented on build-tests/api-documenter-test/build.js:
```diff
@@ -26,7 +26,7 @@ if (process.argv.indexOf('--production') >= 0) {
 // Run the API Documenter command-line
 executeCommand('node node_modules/@microsoft/api-documenter/lib/start '
-  + 'yaml --input-folder etc --output-folder etc/yaml');
+  + 'generate --input-folder etc --output-folder etc/yaml');
```
>generate [](start = 5, length = 8)

Should this be an entirely separate test?

Vitalius1 replied:
The new `generate` action is meant to replace the `yaml` action. It's still under experimental title but as we polish it the plan is having just one action `generate` that will look for an `api-documenter.json` config file where we would specify what documenter to be used: yaml vs markdown.

iclanton commented on apps/api-documenter/src/cli/GenerateAction.ts:
```diff
@@ -0,0 +1,51 @@
+export class GenerateAction extends BaseAction {
+  constructor(parser: ApiDocumenterCommandLine) {
+    super({
+      actionName: 'generate',
```
>generate [](start = 19, length = 8)

This seems like a weird verb. Maybe "experimental-yaml" or something?

Vitalius1 replied:
See [this comment](https://github.com/microsoft/web-build-tools/pull/1265#discussion_r284569163). We discussed it with @octogonz that we might wanna move to a using this generic `generate` action along with a config file that will command which documenter to be used.

Vitalius1 commented on build-tests/api-documenter-test/etc/api-documenter-test.api.json:
```diff
@@ -17,7 +17,7 @@
           "kind": "Class",
           "canonicalReference": "(DocBaseClass:class)",
-          "docComment": "/**\n * Example base class\n *\n * @public\n */\n",
+          "docComment": "/**\n * Example base class\n *\n * {@docCategory DocBaseClass}\n *\n * @public\n */\n",
```
@iclanton I bet this change to the documentation is causing the addition of a blank line. `MarkdownDocumenter` does not know how to handle inline custom tags and it skips `{@docCategory DocBaseClass}` and is left with `\n` right after. So this is specific to only this case and this test project. I don't consider it as a regression. Perhaps a `MarkdownDocumenter` bug that needs to be fixed in a separate PR.

iclanton replied:
Cool.
